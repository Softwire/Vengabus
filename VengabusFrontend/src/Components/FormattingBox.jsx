import React, { Component } from 'react';
import { css } from 'react-emotion';
import { Alert } from 'react-bootstrap';
import formatJSon from 'prettyprint';
const formatXML = require("xml-formatter");


export class FormattingBox extends Component {

    constructor(props) {
        super(props);
        this.state = {
            //for future use if other formats are supported and override is wanted
            isFormattable: true
        }
    }

    render() {

        let formattedText;
        const originalData = this.props.data;
        let xmlFormattingSucceededButChangedText = false;
        let formattingError;
        let mightBeXml, mightBeJson;
        if (originalData[0] === "<" && originalData[originalData.length - 1] === ">") {
            mightBeXml = true;
        }
        if ((originalData[0] === "{" && originalData[originalData.length - 1] === "}") ||
            (originalData[0] === "[" && originalData[originalData.length - 1] === "]")) {
            mightBeJson = true;
        }


        //the XML library returns undefined for not XML meaning that format text will be falsely hence this working 
        if (this.state.isFormattable) {
            try {
                //check for xml first then check for json
                if (mightBeXml) {
                    formattedText = formatXML(originalData);
                    if (formattedText && (formattedText.replace(/\s/g, "") !== originalData)) {
                        xmlFormattingSucceededButChangedText = true;
                    }
                }
                if (!formattedText && mightBeJson) {
                    formattedText = formatJSon(JSON.parse(originalData));
                }
            }
            catch (err) {
                formattingError = err;
            }
        }
        const formatCss = css`
            text-align: left;
        `;
        const xmlChangeAlert = (
            <Alert bsStyle="danger">
                <p>The XML formatter changed the text of this data. This was probably just to 'heal' malformed XML, but we can't be certain.</p>
                <p> See below for the original data text.</p >
            </Alert >
        );
        const errorAlert = (
            <Alert bsStyle="danger">
                <p>{`The formatter threw an error whilst trying to format the text of this data: '${formattingError}'`}</p>
            </Alert>
        );
        const boxContainingOriginalText = (
            <div id = "original">
                <p>Original Data:</p>
                <pre id="originalText" className={formatCss}>
                    {originalData}
                </pre>
            </div>
        )
        const boxContainingFormattedText = (
            <div id = "formatted">
                <p>Formatted Data:</p>
                <pre id="formattedText"className={formatCss}>
                    {formattedText || originalData}
                </pre>
            </div>
        )

        return (
            <div >
                {xmlFormattingSucceededButChangedText ? xmlChangeAlert : ' '}
                {formattingError ? errorAlert : ' '}
                {formattedText ? boxContainingFormattedText : ''}
                {(!formattedText || xmlFormattingSucceededButChangedText) ? boxContainingOriginalText : ' '}
            </div>
        );
    }
}
