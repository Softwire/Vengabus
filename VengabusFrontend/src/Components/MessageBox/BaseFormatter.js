export class BaseFormatter {
    startsAndEndsWith = (inputString, startCharacter, endCharacter) => {
        let noInitialWhitespace = inputString.replace(/^\s*/g, "");
        let noWrappingWhitespace = noInitialWhitespace.replace(/\s*$/g, "");

        return this.startsAndEndsWith_NotIgnoringWhitespace(noWrappingWhitespace, startCharacter, endCharacter);
    }

    startsAndEndsWith_NotIgnoringWhitespace = (inputString, startCharacter, endCharacter) => {
        if (!inputString) {
            return false;
        }
        if (inputString[0] === startCharacter && inputString[inputString.length - 1] === endCharacter) {
            return true;
        }
        return false;
    }

    isMessageTooLong = (originalText) => {
        return originalText && originalText.length > 100000;
    }

    removeBlankLines = (text) => {
        if (!text) {
            return text;
        }
        return text.replace(/^\s*\n/gm, "");
    };

    matchWithoutWhitespaceAndQuotes = (text1, text2) => {
        //ignore quote marks as the JSON formatter sometimes removes these
        return text1.replace(/\s|'|"|`/g, "") === text2.replace(/\s|'|"|`/g, "");
    }

    getConfidenceMatch = (formattedObj, mightBeThisFormat) => {
        //return a float between 0 and 1 indicating how confident we are that the message is of the given format
        if (mightBeThisFormat && !formattedObj.errorMessage) {
            return formattedObj.warningMessage ? 0.5 : 0.9;
        }
        return 0;
    }

    formatResult = (originalText, title, formattingFunction, mightBeThisFormat) => {
        let formattedObj = { formatType: title };
        formattedObj.isMessageTooLongToFormat = this.isMessageTooLong(originalText);
        if (formattedObj.isMessageTooLongToFormat) {
            formattedObj.errorMessage = 'Long message: only messages under 100,000 characters in length are formatted.';
        } else {
            try {
                formattedObj.formattedText = formattingFunction(originalText);
                if (formattedObj.formattedText && !this.matchWithoutWhitespaceAndQuotes(formattedObj.formattedText, originalText)) {
                    formattedObj.warningMessage = `The ${title} formatter changed the text of this data. This was probably just to 'heal' malformed ${title}, but we can't be certain.`;
                }
            }
            catch (error) {
                formattedObj.errorMessage = error.toString();
            }
        }
        formattedObj.matchConfidence = this.getConfidenceMatch(formattedObj, mightBeThisFormat);
        return formattedObj;
    }
}