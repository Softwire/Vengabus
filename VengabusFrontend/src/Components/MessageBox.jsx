import React, { Component } from 'react';
import { Modal, Button } from 'react-bootstrap';
import { FormattingBox } from './FormattingBox';
import { CopyTextButton } from './CopyTextButton';
import { css } from 'emotion';
import { CollapsiblePanel } from './CollapsiblePanel';

export class MessageBox extends Component {

    convertMessagePropertiesToJSXArray = (propertiesArray) => {
        const propertiesJSX = [];
        for (let propName in propertiesArray) {
            let currentPropValue = propertiesArray[propName];
            if (propName !== 'messageId') {
                propertiesJSX.push(
                    <p key={propName}>{propName + ': ' + String(currentPropValue)}</p> /*string conversion needed to display booleans properly*/
                );
            }
        }
        return propertiesJSX.length > 0 ? propertiesJSX : null;
    }

    render() {
        const buttonStyle = css`
            float: right;
            margin-left: 5px;
        `;
        const headerStyle = css`
            font-weight: bold;
            display: inline-block;
        `;
        const messageIdStyle = css`
            display: inline-block;
        `;
        const panelStyle = css`
            .panel {
                margin-bottom: 15px;
            }
        `;
        const message = this.props.message;
        if (!message) {
            return null;
        }
        const preDefinedPropsJSX = this.convertMessagePropertiesToJSXArray(message.predefinedProperties) || <p>There are no pre-defined properties to display</p>;
        const customPropsJSX = this.convertMessagePropertiesToJSXArray(message.customProperties) || <p>There are no user-defined properties to display</p>;
        return (
            <div className="static-modal" >
                <Modal show={this.props.show} onHide={this.props.handleClose} >
                    <Modal.Header>
                        <Modal.Title className={headerStyle}>Message Id:&nbsp;</Modal.Title>
                        <Modal.Title className={messageIdStyle}>{message.predefinedProperties.messageId}</Modal.Title>
                    </Modal.Header>

                    <Modal.Body className={panelStyle}>
                        <CollapsiblePanel panelTitle={"Pre-defined Properties"}>
                            <pre>{preDefinedPropsJSX}</pre>
                        </CollapsiblePanel>
                        <CollapsiblePanel panelTitle={"User-defined Properties"}>
                            <pre>{customPropsJSX}</pre>
                        </CollapsiblePanel>
                        <FormattingBox
                            data={message.messageBody}
                        />
                    </Modal.Body>

                    <Modal.Footer>
                        <Button className={buttonStyle} onClick={this.props.handleClose}>Close</Button>
                        <CopyTextButton className={buttonStyle} text={this.props.messageBody} />
                    </Modal.Footer>
                </Modal>
            </div>

        );
    }
}