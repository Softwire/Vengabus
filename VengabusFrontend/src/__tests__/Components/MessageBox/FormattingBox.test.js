import { FormattingBox } from '../../../Components/MessageBox/FormattingBox';
import renderer from 'react-test-renderer';
import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

function expectFormattedResult(inputMessage, expectedOutput, formatId, expectedDefaultActiveTab) {
    const formattingBox = shallow(<FormattingBox message={inputMessage} />);
    expect(formattingBox.find(formatId)).toExistOnPage();
    const formattedText = formattingBox.find(formatId);
    expect(formattedText.text()).toBe(expectedOutput);
    expect(formattingBox.children().prop('defaultActiveKey')).toBe(expectedDefaultActiveTab);
}

function expectOriginalOutput(inputMessage) {
    expectFormattedResult(inputMessage, inputMessage, '#Original', 0);
}

function expectXMLoutput(inputMessage, expectedOutput) {
    expectFormattedResult(inputMessage, expectedOutput, '#XML', 1);
}

function expectJSONoutput(inputMessage, expectedOutput) {
    expectFormattedResult(inputMessage, expectedOutput, '#JSON', 2);
}

function expectError(inputMessage, errorId) {
    const formattingBox = shallow(<FormattingBox message={inputMessage} />);
    const errorMessage = formattingBox.find(errorId);
    expect(errorMessage).toExistOnPage();
    return errorMessage;
}

describe('FormattingBox', () => {

    describe('renders a snapshot correctly when given', () => {
        it('props', () => {
            let formattingBox = renderer.create(
                <FormattingBox
                    message="test"
                />);
            expect(formattingBox.toJSON()).toMatchSnapshot();
        });

        it('xml', () => {
            let formattingBox = renderer.create(
                <FormattingBox
                    message="<a>This is a sample xml message</a>"
                />);
            expect(formattingBox.toJSON()).toMatchSnapshot();
        });

        it('json', () => {
            let formattingBox = renderer.create(
                <FormattingBox
                    message='{"a":"apple","b":42, "c":false}'
                />);
            expect(formattingBox.toJSON()).toMatchSnapshot();
        });

        it('plain text', () => {
            let formattingBox = renderer.create(
                <FormattingBox
                    message='"a":"apple","b":42, "c":false'
                />);
            expect(formattingBox.toJSON()).toMatchSnapshot();
        });
    });

    describe('guesses the correct format by default', () => {
        it('plain text', () => {
            let formattingBox = shallow(
                <FormattingBox
                    message='"a":"apple","b":42, "c":false'
                />);
            expect(formattingBox.find('[aria-selected=true]')).toMatchSnapshot();
        });

        it('xml', () => {
            let formattingBox = shallow(
                <FormattingBox
                    message="<a>This is a sample xml message</a>"
                />);
            expect(formattingBox.find('[aria-selected=true]')).toMatchSnapshot();
        });

        it('brace json', () => {
            let formattingBox = shallow(
                <FormattingBox
                    message='{"a":"apple","b":42, "c":false}'
                />);
            expect(formattingBox.find('[aria-selected=true]')).toMatchSnapshot();
        });

        it('bracket json', () => {
            let formattingBox = shallow(
                <FormattingBox
                    message='["a":"apple","b":42, "c":false]'
                />);
            expect(formattingBox.find('[aria-selected=true]')).toMatchSnapshot();
        });

    });

    it('does not change plain text', () => {
        const plainText = 'one kilogram of fish is worth 42';
        expectOriginalOutput(plainText);
    });

    it('does not change plain text with line breaks', () => {
        const plainText = `one kilogram
of fish
is worth 42`;
        expectOriginalOutput(plainText);
    });

    it('formats general XML', () => {
        const xmlInput = "<apple><b></b><c>dsagsd{}</c></apple>";
        const expectedOutput =
            `<apple>
    <b>
    </b>
    <c>
        dsagsd{}
    </c>
</apple>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

    it('it Formatted XML with real spaces in', () => {
        const xmlInput = "<apple><b>Some random sentence</b></apple>";
        const expectedOutput =
            `<apple>
    <b>
        Some random sentence
    </b>
</apple>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

    it('formats XML with linebreaks in sentences', () => {
        const xmlInput = `<apple><b>Some random sentence
with linebreaks</b></apple>`;
        const expectedOutput =
            `<apple>
    <b>
        Some random sentence with linebreaks
    </b>
</apple>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

    it('formats XML with tabs in text', () => {
        const xmlInput = `<apple><b>Some random sentence with \t tabs in text</b></apple>`;
        const expectedOutput =
            `<apple>
    <b>
        Some random sentence with \t tabs in text
    </b>
</apple>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

    it('formats XML with tabs between tags', () => {
        const xmlInput = `<apple>\t<b>Some indented tag</b></apple>`;
        const expectedOutput =
            `<apple>
    <b>
        Some indented tag
    </b>
</apple>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

    it('formats XML that had extra spaces between tags', () => {
        const xmlInput = `<apple>    <b>    </b>    <c>#        dsagsd{}    #</c></apple>`;
        const expectedOutput =
            `<apple>
    <b>
    </b>
    <c>
        #        dsagsd{}    #
    </c>
</apple>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

    it('formats XML that was already correctly formatted', () => {
        const xmlInput = `<apple>
    <b>
    </b>
    <c>
        dsa gsd
    </c>
</apple>`;
        expectXMLoutput(xmlInput, xmlInput);
    });

    it('formats simple JSON object', () => {
        const jsonInput = '{"food":"fish", "price":42, "unit":"kilogram"}';
        const expectedOutput =
            `{
    "food": "fish",
    "price": 42,
    "unit": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats simple JSON object with complex keys', () => {
        const jsonInput = '{"food-test":"fish", "price test":42, "unit.test":"kilogram"}';
        const expectedOutput =
            `{
    "food-test": "fish",
    "price test": 42,
    "unit.test": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats simple JSON array', () => {
        const jsonInput = '["fish",42,"kilogram"]';
        const expectedOutput =
            `[
    "fish",
    42,
    "kilogram"
]`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats nested JSON object', () => {
        const jsonInput = '{"food":"fish", "price":{"GBP":42, "USD":54}, "unit":"kilogram"}';
        const expectedOutput =
            `{
    "food": "fish",
    "price": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats mixed JSON array', () => {
        const jsonInput = '["fish",{"GBP":42, "USD":54},"kilogram",false]';
        const expectedOutput =
            `[
    "fish",
    {
        "GBP": 42,
        "USD": 54
    },
    "kilogram",
    false
]`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats JSON that is already formatted', () => {
        const jsonInput = `{
            "food": "fish",
            "price": {
                "GBP": 42,
                "USD": 54
        },
            "unit": "kilogram"
    }`;
        const expectedOutput =
            `{
    "food": "fish",
    "price": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats JSON with linebreaks in sentences', () => {
        const jsonInput = `{
    "fo
    od": "fis
    h",
    "pr\nice": {
        "GBP": 42,

        "USD": 54
    }
    ,

    "unit"
    :
     "kilogram"
}`;
        const expectedOutput =
            `{
    "fo od": "fis h",
    "pr ice": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats JSON that had extra spaces between tags', () => {
        const jsonInput =
            `     {
           "food"    :     "fi  sh",
    "pr  ice": {
        "GBP": 42,
        "USD": 54
    }     ,   
    "unit": "kilogram"    
       }    `;
        const expectedOutput =
            `{
    "food": "fi  sh",
    "pr  ice": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('formats JSON with tabs not inside strings', () => {
        const jsonInput = `\t{\t
    "food"\t: "fish",
    "price": {
        \t"GBP":\t 42\t,\t
        "USD"\t: 54\t
    }\t,\t
    "unit":\t "kilogram"
}\t`;
        const expectedOutput =
            `{
    "food": "fish",
    "price": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilogram"
}`;
        expectJSONoutput(jsonInput, expectedOutput);
    });

    it('fails to format JSON with tabs inside keys', () => {
        //qq because having tabs in keys is not valid JSON. See https://stackoverflow.com/questions/19799006/unable-to-parse-tab-in-json-files
        const jsonInput =
            `{
    "fo\tod": "fish",
    "price": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilogram"
}`;
        const JSONerror = expectError(jsonInput, '#JSONerror');
        expect(toJson(JSONerror)).toMatchSnapshot();
    });

    it('fails to format JSON with tabs inside values', () => {
        //qq because having tabs inside values not valid JSON. See above.
        const jsonInput =
            `{
    "food": "fish",
    "price": {
        "GBP": 42,
        "USD": 54
    },
    "unit": "kilo\tgram"
}`;
        const JSONerror = expectError(jsonInput, '#JSONerror');
        expect(toJson(JSONerror)).toMatchSnapshot();
    });

    it('should fail to format mal-formatted JSON', () => {
        const jsonInput = '{"fish",42,"kilogram"}';
        const JSONerror = expectError(jsonInput, '#JSONerror');
        expect(toJson(JSONerror)).toMatchSnapshot();
    });

    it('should fail to format mal-formatted xml', () => {
        const xmlInput = '<a>fish</a><b>42</b>';
        const XMLerror = expectError(xmlInput, '#XMLerror');
        expect(toJson(XMLerror)).toMatchSnapshot();
    });

    it('heals xml with minor format errors', () => {
        const xmlInput = '<c><d>fish=42</c>';
        const expectedOutput =
            `<c>
    <d>
        fish=42
    </d>
</c>`;
        expectXMLoutput(xmlInput, expectedOutput);
    });

});
