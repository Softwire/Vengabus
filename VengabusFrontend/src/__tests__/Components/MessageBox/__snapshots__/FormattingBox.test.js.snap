// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattingBox renders a snapshot correctly when given json 1`] = `
.emotion-1 {
  white-space: pre-wrap;
}

.emotion-0 {
  margin-left: 5px;
}

.emotion-2 .nav-tabs {
  margin-bottom: 15px;
}

.emotion-2 .tab-content pre {
  margin-bottom: 0;
  overflow: auto;
  max-height: calc(92vh - 326px);
}

.emotion-2 .formatterError a {
  color: #a94442;
  background: #f2dede;
}

.emotion-2 .formatterError.active>a,
.emotion-2 .formatterError.active>a:focus,
.emotion-2 .formatterError.active>a:hover {
  color: #a94442;
}

.emotion-2 .formatterWarning a {
  color: #8a6d3b;
  background: #fcf8e3;
}

.emotion-2 .formatterWarning.active>a,
.emotion-2 .formatterWarning.active>a:focus,
.emotion-2 .formatterWarning.active>a:hover {
  color: #8a6d3b;
}

<div>
  <div
    className="emotion-2"
    id="message-formatting-tabs"
    style={undefined}
  >
    <ul
      className="nav nav-tabs"
      role="tablist"
    >
      <li
        className=""
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-0"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Original
        </a>
      </li>
      <li
        className="formatterError"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-1"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-1"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          XML
          <span
            className="emotion-0 glyphicon glyphicon-exclamation-sign"
          />
        </a>
      </li>
      <li
        className="active"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-2"
          aria-selected={true}
          href="#"
          id="message-formatting-tabs-tab-2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={undefined}
        >
          JSON
        </a>
      </li>
    </ul>
    <div
      className="tab-content"
    >
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-0"
        className="tab-pane"
        id="message-formatting-tabs-pane-0"
        role="tabpanel"
      >
        <pre
          className="emotion-1"
          id="Original"
        >
          {"a":"apple","b":42, "c":false}
        </pre>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-1"
        className="tab-pane"
        id="message-formatting-tabs-pane-1"
        role="tabpanel"
      >
        <div
          className="alert alert-danger"
          id="XMLwarning"
          role="alert"
        >
          <p>
            The formatter didn't return any text to display.
          </p>
        </div>
      </div>
      <div
        aria-hidden={false}
        aria-labelledby="message-formatting-tabs-tab-2"
        className="tab-pane active"
        id="message-formatting-tabs-pane-2"
        role="tabpanel"
      >
        <pre
          className=""
          id="JSON"
        >
          {
              "a": "apple",
              "b": 42,
              "c": false
          }
        </pre>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormattingBox renders a snapshot correctly when given plain text 1`] = `
.emotion-2 {
  white-space: pre-wrap;
}

.emotion-0 {
  margin-left: 5px;
}

.emotion-3 .nav-tabs {
  margin-bottom: 15px;
}

.emotion-3 .tab-content pre {
  margin-bottom: 0;
  overflow: auto;
  max-height: calc(92vh - 326px);
}

.emotion-3 .formatterError a {
  color: #a94442;
  background: #f2dede;
}

.emotion-3 .formatterError.active>a,
.emotion-3 .formatterError.active>a:focus,
.emotion-3 .formatterError.active>a:hover {
  color: #a94442;
}

.emotion-3 .formatterWarning a {
  color: #8a6d3b;
  background: #fcf8e3;
}

.emotion-3 .formatterWarning.active>a,
.emotion-3 .formatterWarning.active>a:focus,
.emotion-3 .formatterWarning.active>a:hover {
  color: #8a6d3b;
}

<div>
  <div
    className="emotion-3"
    id="message-formatting-tabs"
    style={undefined}
  >
    <ul
      className="nav nav-tabs"
      role="tablist"
    >
      <li
        className="active"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-0"
          aria-selected={true}
          href="#"
          id="message-formatting-tabs-tab-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={undefined}
        >
          Original
        </a>
      </li>
      <li
        className="formatterError"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-1"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-1"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          XML
          <span
            className="emotion-0 glyphicon glyphicon-exclamation-sign"
          />
        </a>
      </li>
      <li
        className="formatterError"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-2"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          JSON
          <span
            className="emotion-0 glyphicon glyphicon-exclamation-sign"
          />
        </a>
      </li>
    </ul>
    <div
      className="tab-content"
    >
      <div
        aria-hidden={false}
        aria-labelledby="message-formatting-tabs-tab-0"
        className="tab-pane active"
        id="message-formatting-tabs-pane-0"
        role="tabpanel"
      >
        <pre
          className="emotion-2"
          id="Original"
        >
          "a":"apple","b":42, "c":false
        </pre>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-1"
        className="tab-pane"
        id="message-formatting-tabs-pane-1"
        role="tabpanel"
      >
        <div
          className="alert alert-danger"
          id="XMLwarning"
          role="alert"
        >
          <p>
            The formatter didn't return any text to display.
          </p>
        </div>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-2"
        className="tab-pane"
        id="message-formatting-tabs-pane-2"
        role="tabpanel"
      >
        <div
          className="alert alert-danger"
          id="JSONerror"
          role="alert"
        >
          <p>
            The formatter threw an error whilst trying to format the text of this data:
          </p>
          <p>
            SyntaxError: Unexpected token : in JSON at position 3
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormattingBox renders a snapshot correctly when given props 1`] = `
.emotion-2 {
  white-space: pre-wrap;
}

.emotion-0 {
  margin-left: 5px;
}

.emotion-3 .nav-tabs {
  margin-bottom: 15px;
}

.emotion-3 .tab-content pre {
  margin-bottom: 0;
  overflow: auto;
  max-height: calc(92vh - 326px);
}

.emotion-3 .formatterError a {
  color: #a94442;
  background: #f2dede;
}

.emotion-3 .formatterError.active>a,
.emotion-3 .formatterError.active>a:focus,
.emotion-3 .formatterError.active>a:hover {
  color: #a94442;
}

.emotion-3 .formatterWarning a {
  color: #8a6d3b;
  background: #fcf8e3;
}

.emotion-3 .formatterWarning.active>a,
.emotion-3 .formatterWarning.active>a:focus,
.emotion-3 .formatterWarning.active>a:hover {
  color: #8a6d3b;
}

<div>
  <div
    className="emotion-3"
    id="message-formatting-tabs"
    style={undefined}
  >
    <ul
      className="nav nav-tabs"
      role="tablist"
    >
      <li
        className="active"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-0"
          aria-selected={true}
          href="#"
          id="message-formatting-tabs-tab-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={undefined}
        >
          Original
        </a>
      </li>
      <li
        className="formatterError"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-1"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-1"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          XML
          <span
            className="emotion-0 glyphicon glyphicon-exclamation-sign"
          />
        </a>
      </li>
      <li
        className="formatterError"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-2"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          JSON
          <span
            className="emotion-0 glyphicon glyphicon-exclamation-sign"
          />
        </a>
      </li>
    </ul>
    <div
      className="tab-content"
    >
      <div
        aria-hidden={false}
        aria-labelledby="message-formatting-tabs-tab-0"
        className="tab-pane active"
        id="message-formatting-tabs-pane-0"
        role="tabpanel"
      >
        <pre
          className="emotion-2"
          id="Original"
        >
          test
        </pre>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-1"
        className="tab-pane"
        id="message-formatting-tabs-pane-1"
        role="tabpanel"
      >
        <div
          className="alert alert-danger"
          id="XMLwarning"
          role="alert"
        >
          <p>
            The formatter didn't return any text to display.
          </p>
        </div>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-2"
        className="tab-pane"
        id="message-formatting-tabs-pane-2"
        role="tabpanel"
      >
        <div
          className="alert alert-danger"
          id="JSONerror"
          role="alert"
        >
          <p>
            The formatter threw an error whilst trying to format the text of this data:
          </p>
          <p>
            SyntaxError: Unexpected token e in JSON at position 1
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormattingBox renders a snapshot correctly when given xml 1`] = `
.emotion-1 {
  white-space: pre-wrap;
}

.emotion-0 {
  margin-left: 5px;
}

.emotion-2 .nav-tabs {
  margin-bottom: 15px;
}

.emotion-2 .tab-content pre {
  margin-bottom: 0;
  overflow: auto;
  max-height: calc(92vh - 326px);
}

.emotion-2 .formatterError a {
  color: #a94442;
  background: #f2dede;
}

.emotion-2 .formatterError.active>a,
.emotion-2 .formatterError.active>a:focus,
.emotion-2 .formatterError.active>a:hover {
  color: #a94442;
}

.emotion-2 .formatterWarning a {
  color: #8a6d3b;
  background: #fcf8e3;
}

.emotion-2 .formatterWarning.active>a,
.emotion-2 .formatterWarning.active>a:focus,
.emotion-2 .formatterWarning.active>a:hover {
  color: #8a6d3b;
}

<div>
  <div
    className="emotion-2"
    id="message-formatting-tabs"
    style={undefined}
  >
    <ul
      className="nav nav-tabs"
      role="tablist"
    >
      <li
        className=""
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-0"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Original
        </a>
      </li>
      <li
        className="active"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-1"
          aria-selected={true}
          href="#"
          id="message-formatting-tabs-tab-1"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={undefined}
        >
          XML
        </a>
      </li>
      <li
        className="formatterError"
        role="presentation"
        style={undefined}
      >
        <a
          aria-controls="message-formatting-tabs-pane-2"
          aria-selected={false}
          href="#"
          id="message-formatting-tabs-tab-2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="tab"
          tabIndex={-1}
        >
          JSON
          <span
            className="emotion-0 glyphicon glyphicon-exclamation-sign"
          />
        </a>
      </li>
    </ul>
    <div
      className="tab-content"
    >
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-0"
        className="tab-pane"
        id="message-formatting-tabs-pane-0"
        role="tabpanel"
      >
        <pre
          className="emotion-1"
          id="Original"
        >
          &lt;a&gt;This is a sample xml message&lt;/a&gt;
        </pre>
      </div>
      <div
        aria-hidden={false}
        aria-labelledby="message-formatting-tabs-tab-1"
        className="tab-pane active"
        id="message-formatting-tabs-pane-1"
        role="tabpanel"
      >
        <pre
          className=""
          id="XML"
        >
          &lt;a&gt;
              This is a sample xml message
          &lt;/a&gt;
        </pre>
      </div>
      <div
        aria-hidden={true}
        aria-labelledby="message-formatting-tabs-tab-2"
        className="tab-pane"
        id="message-formatting-tabs-pane-2"
        role="tabpanel"
      >
        <div
          className="alert alert-danger"
          id="JSONerror"
          role="alert"
        >
          <p>
            The formatter threw an error whilst trying to format the text of this data:
          </p>
          <p>
            SyntaxError: Unexpected token &lt; in JSON at position 0
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FormattingBox should fail to format mal-formatted JSON 1`] = `
<Alert
  bsClass="alert"
  bsStyle="danger"
  closeLabel="Close alert"
  id="JSONerror"
  key="JSONerror"
>
  <p
    key="0"
  >
    The formatter threw an error whilst trying to format the text of this data:
  </p>
  <p
    key="1"
  >
    SyntaxError: Unexpected token , in JSON at position 7
  </p>
</Alert>
`;

exports[`FormattingBox should fail to format mal-formatted xml 1`] = `
<Alert
  bsClass="alert"
  bsStyle="danger"
  closeLabel="Close alert"
  id="XMLerror"
  key="XMLerror"
>
  <p
    key="0"
  >
    The formatter threw an error whilst trying to format the text of this data:
  </p>
  <p
    key="1"
  >
    Error: Found multiple root nodes
  </p>
</Alert>
`;
